src/Shared/PackageManagers/BaseProjectManager.cs

```diff
@@ -310,6 +310,19 @@ public virtual Task<IEnumerable<string>> EnumerateVersionsAsync(PackageURL purl,
         {
             throw new NotImplementedException("BaseProjectManager does not implement EnumerateVersions.");
         }
+        
+        /// <summary>
+        /// Enumerates all possible versions of the package identified by purl, in descending order with the time they were published at.
+        /// </summary>
+        /// <remarks>The latest version is always first, then it is sorted by SemVer in descending order.</remarks>
+        /// <param name="purl">Package URL specifying the package. Version is ignored.</param>
+        /// <param name="useCache">If the cache should be used when looking for the versions.</param>
+        /// <param name="includePrerelease">If pre-release versions should be included.</param>
+        /// <returns> A dictionary where the key is the package version, and the value is the time it was published at.</returns>
+        public virtual Task<IDictionary<string, DateTime>> EnumerateVersionsWithPublishTimeAsync(PackageURL purl, bool useCache = true, bool includePrerelease = true)
```

Comment:
Do we have any package-type-agnostic use cases in mind for this method at this point? Or could this just be defined on the NPM-specific project manager to start?

We don't want users of the library to frequently stumble on `NotImplementedException` when we haven't implemented base methods like this for many package types yet.

Reply:
Right now it is just npm yes, I just didn't know if it made more sense from our side of things to handle a `NotImplementedException` vs doing a check for it being an NPM package, and then casting the `BaseProjectManager` to `NpmProjectManager` to then call the method.

Reply:
@jpinz if you don’t intend to implement this for other managers it does not belong on baseprojectmanager.

Reply:
I'm removing it from this PR until we can come up with a good solution for getting metadata for all versions of a given package.

Reply:
I just implemented it as an extension method in our project which for the time being is exactly what we need.